### PR TITLE
Make the help text of "pihole checkout [what] [branch]" more colorful

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

See title.

image
---

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

File: pihole [hunk 1]
```diff
@@ -408,19 +408,21 @@ piholeCheckoutFunc() {
...
-
-  Branches:
-    master              Update subsystems to the latest stable release
-    dev                 Update subsystems to the latest development release
-    branchname          Update subsystems to the specified branchname"
+      echo "Switch Pi-hole subsystems to a different GitHub branch
+    Usage: ${COL_GREEN}pihole checkout${COL_NC} ${COL_YELLOW}shortcut${COL_NC}
+        or ${COL_GREEN}pihole checkout${COL_NC} ${COL_PURPLE}repo${COL_NC} ${COL_CYAN}branch${COL_NC}
+  Example: ${COL_GREEN}pihole checkout${COL_NC} ${COL_YELLOW}master${COL_NC}
+       or  ${COL_GREEN}pihole checkout${COL_NC} ${COL_PURPLE}ftl ${COL_CYAN}development${COL_NC}
```
Comment (rdwebdesign) [inline pihole:415]: The "Usage/Example" block is a little confusing.

Maybe adding a blank line and changing the indentation will look beter:

```suggestion
  Usage:
    ${COL_GREEN}pihole checkout${COL_NC} ${COL_YELLOW}shortcut${COL_NC}
    ${COL_GREEN}pihole checkout${COL_NC} ${COL_PURPLE}repo${COL_NC} ${COL_CYAN}branch${COL_NC}

  Example:
    ${COL_GREEN}pihole checkout${COL_NC} ${COL_YELLOW}master${COL_NC}
    ${COL_GREEN}pihole checkout${COL_NC} ${COL_PURPLE}ftl ${COL_CYAN}development${COL_NC}
```

image

**Review by rdwebdesign — CHANGES_REQUESTED** [review]

**Review by yubiuser — CHANGES_REQUESTED** [review]
Please do not combine these two unrelated changes:
1) Changing the color -> fine
2) adding a new `Sourcing` variable for `basic-install.sh`

Both changes are not related at all. Additionally, we already have a variable indicating "sourcing" of the installer script: `SKIP_INSTALL` ;-)

**Comment by DL6ER** [comment]
It is not unrelated at all, please see the comment I added above the new `if` in `basic-install.sh`. `SKIP_INSTALL` does not prevent `set -e` and extending `SKIP_INSTALL` to do `set -e` may have unwanted side-effects in other scripts.

The point is that `pihole checkout ftl non/existing_branch` is currently entirely broken. As soon as the script detects that a branch is not available, it does a `return 1` internally and `set -e` kills the script **instead** of the script continuing and showing a list of available branches as it ought to be.

Hence, we indeed need both changes in exactly this combination or you cannot review the changes properly. As the variable is only added in one place and does not exist anywhere else, there are no side-effects.

**Comment by yubiuser** [comment]
I see what you mean, but don't think this is the best solution. The issue I see here is that the whole check for available FTL binaries/branches is mixed and reverted compared to `core` and `web`. In the latter you first get the list of available branches and then check if the desired branch is part of the array. For FTL you check if the binary server has the directory and if not, you check the branches on github. Those two should match in theory, but this might not always be the case. 

For the current situation I see a very simple solution wich does not require a new variable (I don't like the entanglement of all those bash scripts) and a more complex solution.
Simple is to add `set +e` in the checkout script after `basic-install`has been sourced. 
The complex is getting a list of directories from `ftl.pihole.net` and check if the desired path is within this array. Something like 

```
curl --silent 'https://ftl.pi-hole.net/?'  --data-raw '{"action":"get","items":{"href":"/","what":1}}' | jq --raw-output .items[].href
```

**Comment by DL6ER** [comment]
I think this is a well-justified difference between the repos. You can really only checkout what exists as a branch for the two repos but FTL does not really work on a branch-basis (of course, but it does so in a rather indirect way)

I don't really like the proposed idea of getting a list of downloadable stuff as it'd require at least maintaining a filter list to excluding possible library mirror paths we have on there, plus what might be added in the future for similar reasons.

First querying GitHub for the availability of branches needs another step and, hence, was probably not considered for speed/efficiency/need to actually do it this way. It just doesn't matter if the branch exists or not as long as the binary is there. 

Thinking about it, it's nonetheless probably better to first check for the branch availability as it will prevent users finding old topics having some `pihole checkout ftl ...` command for something that has long been merged (= branch deleted) but where the corresponding binaries may still exist due to their significantly longer retention times.

I undid the new variable change to `basic-install.sh` and fixed it by other means. Currently, `check_download_exists()` (and how it is used) is broken in `development-v6`.

**Review by rdwebdesign — APPROVED** [review]
From the formatting/visual point of view it is approved.

**Review by yubiuser — APPROVED** [review]
```
  [i] Checking for availability of branch fix/teleporter_v5_cnames on GitHub
  [✓] Branch fix/teleporter_v5_cnames exists on GitHub
  [i] Checking for pihole-FTL-arm64 binary on https://ftl.pi-hole.net
  [✗] Binary for requested branch is not available, please try again later.
      If the issue persists, please contact Pi-hole Support and ask them to re-generate the binary.

```
```
  [i] Checking for availability of branch asdad on GitHub
  [✗] Requested branch is not available

  [i] Available branches are:
      - ci/macos14
      - development
      - development-v6
      - fix/cli_pw_restart
      - fix/ede_neterr
...
```

**Pull request merged**